iam-token-manager/v1.ts

```diff
@@ -0,0 +1,190 @@
+   */
+  public getManagedToken(cb: Function) {
+    if (this.userAccessToken) {
+      // 1. use user-managed token
+      return cb(this.userAccessToken);
+    } else if (this.needToRequestToken()) {
+      // 2. request an initial token
+      this.requestToken({}, tokenResponse => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(this.tokenInfo.access_token);
+      });
+    } else if (this.tokenIsExpired()) {
+      // 3. refresh a token
+      this.refreshToken({}, tokenResponse => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(this.tokenInfo.access_token);
+      });
+    } else {
+      // 4. use valid managed token
+      return cb(this.tokenInfo.access_token);
+    }
+  }
+
+  public requestToken(params: TokenRequestParams, cb: Function): void {
+    const contentType = params.contentType ||
+      'application/x-www-form-urlencoded';
+
+    const grantType = params.grantType ||
+      'urn:ibm:params:oauth:grant-type:apikey';
+
+    const authorization = params.authHeader || 'Basic Yng6Yng=';
+    const responseType = params.responseType || 'cloud_iam';
+
+    const requestOptions = {
+      url: this.url,
+      headers: {
+        'Content-type': contentType,
+        Authorization: authorization
+      },
+      form: {
+        grant_type: grantType,
+        apikey: this.apikey,
+        response_type: responseType
+      }
+    };
+    request.post(requestOptions, (err, res, body) => {
+      if (err) {
+        throw new Error(err);
+      } else if (res.statusCode < 200 || res.statusCode >= 300) {
+        throw new Error(JSON.stringify(res, null, 2));
+      } else {
+        const tokenResponse = JSON.parse(body);
+        cb(tokenResponse);
+      }
+    });
+  }
+
+  public refreshToken(params: TokenRefreshParams, cb: Function) {
+    const authorization = params.authHeader || 'Basic Yng6Yng=';
+    const contentType = params.contentType ||
+      'application/x-www-form-urlencoded';
+
+    const requestOptions = {
+      url: this.url,
+      headers: {
+        'Content-type': contentType,
+        Authorization: authorization
+      },
+      form: {
+        grant_type: 'refresh_token',
+        refresh_token: this.tokenInfo.refresh_token
+      }
+    };
+
+    request.post(requestOptions, (err, res, body) => {
+      if (err) {
+        throw new Error(err);
+      } else if (res.statusCode < 200 || res.statusCode >= 300) {
+        throw new Error(JSON.stringify(res, null, 2));
+      } else {
+        const tokenResponse = JSON.parse(body);
+        cb(tokenResponse);
+      }
+    });
+  }
+
+  public setAccessToken(access_token: string): void { // tslint:disable-line variable-name
+    this.userAccessToken = access_token;
+  }
+
+  private needToRequestToken() {
+    return !this.tokenInfo.access_token;
+  }
+
+  private tokenIsExpired(): boolean {
+    // use a buffer to prevent the edge case of the token
```

Comment:
Can you move this comment as a method description?

Reply:
yep, good call